plugins/kinetic-dp/fu-kinetic-dp-secure-aux-isp.c

```diff
@@ -0,0 +1,873 @@
+	}
+
+	if (KT_DPCD_STS_SECURE_ENABLED == status) {
+		is_isp_secure_auth_mode = TRUE;
+	} else {
+		is_isp_secure_auth_mode = FALSE;
+		isp_total_data_size -= (FW_CERTIFICATE_SIZE * 2 + FW_RSA_SIGNATURE_BLOCK_SIZE * 2);
+	}
+
+	if (!fu_kinetic_dp_secure_aux_isp_read_dpcd_reply_data_reg (connection, reply_data, sizeof (reply_data), &read_len, error)) {
+		g_prefix_error (error, "Failed to read flash ID and size: ");
+		return FALSE;
+	}
+
+	flash_id = (guint16)reply_data[0] << 8 | reply_data[1];
+	flash_size = (guint16)reply_data[2] << 8 | reply_data[3];
+	read_flash_prog_time = (guint16)reply_data[4] << 8 | reply_data[5];
+
+	if (read_flash_prog_time == 0)
+		read_flash_prog_time = 10;
+
+	return TRUE;
+}
+
+static gboolean
+fu_kinetic_dp_secure_aux_isp_send_isp_drv (FuKineticDpConnection *connection,
+					   gboolean is_app_mode,
+					   const guint8 *isp_drv_data,
+					   guint32 isp_drv_len,
+					   GError **error)
+{
+	g_debug ("Sending ISP driver payload... started");
+
+	if (fu_kinetic_dp_secure_aux_isp_enter_code_loading_mode (connection, is_app_mode, isp_drv_len, error) == FALSE) {
+		g_prefix_error (error, "Enabling code-loading mode... failed: ");
+		return FALSE;
+	}
+
+	if (!fu_kinetic_dp_secure_aux_isp_send_payload (connection, isp_drv_data, isp_drv_len, 10000, 50, error)) {
+		g_prefix_error (error, "Sending ISP driver payload... failed: ");
+		return FALSE;
+	}
+
+	g_debug ("Sending ISP driver payload... done!");
+
+	if (!fu_kinetic_dp_secure_aux_isp_execute_isp_drv (connection, error))
+	{
+		g_prefix_error (error, "ISP driver booting up... failed: ");
+		return FALSE;
+	}
+
+	g_debug ("Flash ID: 0x%04X", flash_id);
+
+	if (flash_size) {
+		if (flash_size < 2048)	/* One bank size in Jaguar is 1024KB */
+			g_debug ("Flash Size: %d KB, Dual Bank is not supported!", flash_size);
+		else
+			g_debug ("Flash Size: %d KB", flash_size);
+	} else {
+		if (flash_id)
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "SPI flash not supported");
+		else
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "SPI flash not connected");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_kinetic_dp_secure_aux_isp_enable_fw_update_mode (FuKineticDpFirmware *firmware,
+						    FuKineticDpConnection *connection,
+						    GError **error)
+{
+	guint8 status;
+	guint8 pl_size_data[12] = {0};
+
+	g_debug ("Entering F/W loading mode...");
+
+	/* Send payload size to DPCD_MCA_REPLY_DATA_REG */
+	*(guint32 *)pl_size_data = fu_kinetic_dp_firmware_get_esm_payload_size (firmware);
+	*(guint32 *)&pl_size_data[4] = fu_kinetic_dp_firmware_get_arm_app_code_size (firmware);
+	*(guint16 *)&pl_size_data[8] = (guint16)fu_kinetic_dp_firmware_get_app_init_data_size (firmware);
+	*(guint16 *)&pl_size_data[10] = (guint16)fu_kinetic_dp_firmware_get_cmdb_block_size (firmware) |
+					(fu_kinetic_dp_firmware_get_is_fw_esm_xip_enabled (firmware) << 15);
```

Comment:
Please use fu_common_read_uint32_safe() -- that's not even remotely safe on non-LE systems